### PR TITLE
filescom: add Files.com backend

### PR DESCRIPTION
**Pull request by childish-sambino**
#### What is the purpose of this change?

This change adds a new backend for [Files.com](https://www.files.com/): `filescom`

#### Was the change discussed in an issue or in the forum before?

No.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by ncw** [comment]
Hi!

How are you doing with the integration tests?

Are you following the instructions here? https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend

I had a brief skim of the code - seems to be heading in the right direction - well done :-)

I'm going to need an account for the integration tester which is run daily. Are free ones available? I see from your linked in that you work for files.com so maybe you could arrange a perpetual testing account?

I wonder whether calling the backend `files` is maybe a bit too generic a name and likely to confuse users. Maybe `filescom`? What do you think

Let me know if you need any help - writing an rclone backend isn't easy!

**Comment by childish-sambino** [comment]
@ncw Thanks for taking a look!

Haven't yet started with automated testing yet. I'm following the contributing guide, and I'll move this out of draft once ready for full review. I've updated the name to `filescom`, which we have prior art of using when it makes sense. Definitely not easy writing a new backend, but at least there are plenty of examples to draw inspiration from.

For the testing account, we can do that. We'll likely email you a link to create a user in the account, and then you can create an API key for use with the daily integration tester.

We're using our Files.com Go SDK to handle all the API operations, errors, filename normalization, etc. Hope this is okay, given all the added dependencies (: It greatly reduces the LOC needed by this backend.

**Comment by childish-sambino** [comment]
* Integration tests work most of the time. Debugging a potential backend issue for transient failures.
* Awaiting internal process for granting a test account.
* Now passing in custom `fshttp` client and verified as working properly.
* Looking into documentation ...

**Comment by childish-sambino** [comment]
Continuing to debug a potential backend issue for transient integration test failures and awaiting internal process for granting a test account.

**Comment by childish-sambino** [comment]
Transient integration test failure fixed. Working on getting the transitive dependency updated.

**Comment by ncw** [comment]
Ready for a review?

**Comment by childish-sambino** [comment]
@ncw Yup, ready for 👀

**Comment by childish-sambino** [comment]
@ncw Hey, Nick. I'm still awaiting internal dependencies for getting the test account to you.

The build failure looks unrelated to these changes, but is consistently failing.

I've some additional changes to add session-based auth as an option to this backend, so I'll be pushing changes for that probably today or tomorrow.

Lastly, would you be open to allowing this backend to use an image for the logo instead of a fontawesome icon (both the navbar and the content page)?

**Comment by childish-sambino** [comment]
Test account invite sent.

**Comment by ncw** [comment]
@childish-sambino 
> The build failure looks unrelated to these changes, but is consistently failing.

rclone has tripped over a bug in the Go wasm support. I've disabled it for the moment so if you rebase on master that should be fixed.

> I've some additional changes to add session-based auth as an option to this backend, so I'll be pushing changes for that probably today or tomorrow.

Great
 
> Lastly, would you be open to allowing this backend to use an image for the logo instead of a fontawesome icon (both the navbar and the content page)?

I've avoided doing that in the past as I don't want copyrighted material in the open source repo and I don't want to be hotlinking logos held elsewhere. Also getting permission for logo use is hard so I don't think I'd get many logos on the site.

> Test account invite sent.

Received - thank you.

I will start reviewing this now, but may not finish until tomorrow!

**Comment by childish-sambino** [comment]
Branched synced. Hope the force-push didn't muck up your review process.

File: docs/content/filescom.md
Comment (ncw) [inline docs/content/filescom.md:101]: This seems to be missing the autogenerated markers so these options get filled in automatically.

They look like

```
{{< rem autogenerated options start" - DO NOT EDIT - instead edit fs.RegInfo in backend/dropbox/dropbox.go then run make backenddocs" >}}
{{< rem autogenerated options stop >}}
```

And `bin/make_backend_docs.py filescom` should put them in for you.

File: docs/content/filescom.md
Comment (ncw) [inline docs/content/filescom.md:13]: Can you move these lines to the how to use section below?

They've been copied around from backend docs and I'm trying to get rid of them!

File: docs/content/filescom.md
Comment (ncw) [inline docs/content/filescom.md:15]: I think this should probably say

The initial setup for `filescom` involves

File: docs/content/overview.md
Comment (ncw) [inline docs/content/overview.md:27]: You haven't implemented setting directory mod times so this should just be `R/W`

It is relatively easy to do if you want to

**Review by ncw — COMMENTED** [review]
This looks like a great job - well done :-)

I've put some things which need looking at inline.

If you want to implement setting directory modification times, take a look at the `DirSetModTime` method (eg in the sftp backend) and consider whether you need to set the `DirModTimeUpdatesOnWrite` feature flag.

This has lots of dependencies, which I don't mind provided they don't actually get linked in. Can you check the binary size of rclone with and without the filescom backend?

You could consider making the individual parts of the files.com SDK their own go modules. They still all live in the same repo but each has a go.mod file, you just need to "go get" the parts you need. AWS did this to split their massive SDK up a while back and it seems to have worked quite well and I think the go tooling with workspaces supports this style of working quite well now.

**Comment by childish-sambino** [comment]
Will take a look through the comments and addresses as needed. Just pulled and built the latest w/ and w/o these changes:

```
w/
94,410,482
w/o
89,804,594
diff
4,605,888
```

File: docs/content/overview.md
Comment (childish-sambino) [reply docs/content/overview.md:27]: Added directory mod time setting.

File: backend/filescom/filescom.go
Comment (childish-sambino) [reply backend/filescom/filescom.go]: Ah, missed it. Good catch. Updated.

File: backend/filescom/filescom.go
Comment (childish-sambino) [reply backend/filescom/filescom.go:701]: Both hashes are calculated async, so they could come back as empty.

File: backend/filescom/filescom.go
Comment (childish-sambino) [reply backend/filescom/filescom.go]: Yup, that works. Updated.

File: backend/filescom/filescom.go
Comment (childish-sambino) [reply backend/filescom/filescom.go]: Works for me. Updated.

File: docs/content/overview.md
Comment (childish-sambino) [reply docs/content/overview.md:27]: Actually, still working on this ...

Comment (childish-sambino) [reply docs/content/overview.md:27]: Updated.

File: docs/content/filescom.md
Comment (ncw) [reply docs/content/filescom.md:14]: I should probably change the box backend.

I suggest you just remove those lines to make the docs read better.

**Comment by ncw** [comment]
> Will take a look through the comments and addresses as needed. Just pulled and built the latest w/ and w/o these changes:
> 
> ```
> w/
> 94,410,482
> w/o
> 89,804,594
> diff
> 4,605,888
> ```

More than I would have expected. I wonder what deps it pulls in to the binary? There is a tool to analyse that but I can't remember it's name right now...

**Comment by childish-sambino** [comment]
I'm using https://github.com/Zxilly/go-size-analyzer to see where the bloat is.

I was able to chop the extra in half pretty easily by moving some test utils to test files in the Go SDK. Locally it's now:

```
w/
92,139,762
w/o
89,804,594
diff
2,335,168
```

File: docs/content/filescom.md
Comment (childish-sambino) [reply docs/content/filescom.md]: Updated.

**Comment by ncw** [comment]
> I'm using https://github.com/Zxilly/go-size-analyzer to see where the bloat is.
> 
> I was able to chop the extra in half pretty easily by moving some test utils to test files in the Go SDK. Locally it's now:
> 
> ```
> w/
> 92,139,762
> w/o
> 89,804,594
> diff
> 2,335,168
> ```

That is a good improvement.

I wrote a little tool  [test_backend_sizes.py](https://github.com/rclone/rclone/blob/master/bin/test_backend_sizes.py) to test the effect of compiling rclone with and without a backend. This uses the production build flags which strip debugging and trim the path.

Can you try out `bin/test_backend_sizes.py filescom` and see what you get?

I made an issue about large backends here #7998 and it would be nice if the filescom backend doesn't appear in the top 5!

The backends which use `lib/rest` are usually about 200k

**Comment by childish-sambino** [comment]
Current:
```
❯ bin/test_backend_sizes.py filescom
Total binary size 71.833 MiB
Backend,Size MiB
filescom,3.495
```

With some squeezing in the Files.com Go SDK:
```
❯ bin/test_backend_sizes.py filescom
Total binary size 70.223 MiB
Backend,Size MiB
filescom,1.885
```

**Comment by ncw** [comment]
> With some squeezing in the Files.com Go SDK:
> 
> ```
> ❯ bin/test_backend_sizes.py filescom
> Total binary size 70.223 MiB
> Backend,Size MiB
> filescom,1.885
> ```

That would put it in at number 5 which is a great improvement. Any more squeezing possible?

**Comment by childish-sambino** [comment]
Latest proposed changes drop us out of the top 5:
```
❯ bin/test_backend_sizes.py filescom
Total binary size 70.098 MiB
Backend,Size MiB
filescom,1.760
```

I don't think we're going to get much further below that without some major changes.

**Comment by ncw** [comment]
> Latest proposed changes drop us out of the top 5:
> 
> ```
> ❯ bin/test_backend_sizes.py filescom
> Total binary size 70.098 MiB
> Backend,Size MiB
> filescom,1.760
> ```
> 
> I don't think we're going to get much further below that without some major changes.

OK let's call that good enough. Well done for squeezing it down! I'll just give the code a quick once over after your changes.

**Review by ncw — APPROVED** [review]
This is looking great now - thank you :-)

Two things for you to confirm and then I'll merge

1. Is this is using the slimmed down filescom SDK in the `go.mod`?
2. Is the full suite integration tests still passing after your changes?

Thank you

**Comment by childish-sambino** [comment]
Latest sizes using the latest released Files.com Go SDK:

```
❯ bin/test_backend_sizes.py filescom
Total binary size 70.173 MiB
Backend,Size MiB
filescom,1.760
```

So looking good there. Seeing some failures when running the full integ tests against this backend. Investigating ...

**Comment by childish-sambino** [comment]
Some bisync tests are failing pretty consistently. Will debug in the AM.

**Comment by ncw** [comment]
@childish-sambino not all the bisync tests on all the backends at the moment - that is work in progress! But if you can fix them in the backend - great. However there may be stuff elsewhere which needs fixing.

**Comment by childish-sambino** [comment]
I'm only running the tests against this new backend, so I assume those need to be passing to get this merged.

**Comment by childish-sambino** [comment]
@ncw All integration tests now passing (including bisync) when running `test_all -backends filescom`.

However, I added support for the `link` command and it slightly increased the size of the backend:

```
❯ bin/test_backend_sizes.py filescom
Total binary size 70.269 MiB
Backend,Size MiB
filescom,1.857
```

**Comment by ncw** [comment]
@childish-sambino 
> All integration tests now passing (including bisync) when running `test_all -backends filescom`.

Great - well done :-)

> However, I added support for the `link` command and it slightly increased the size of the backend:

That is fine.

I gave the backend a smoke test and this did not work

```
$ date -Is | rclone rcat TestFilesCom:file.txt
$ rclone cat TestFilesCom:file.txt
2024/08/15 11:32:04 ERROR : : error listing: directory not found
2024/08/15 11:32:04 Failed to cat with 2 errors: last error was: directory not found
```

Which is very surprising since your backend passed all the integration tests.

The fix looks like this (this is a common gotcha with path.Dir in the rclone code)

```diff
--- a/backend/filescom/filescom.go
+++ b/backend/filescom/filescom.go
@@ -235,6 +235,10 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
 		info, err := f.readMetaDataForPath(ctx, "")
 		if err == nil && !info.IsDir() {
 			f.root = path.Dir(f.root)
+			if f.root == "." {
+				f.root = ""
+			}
 			return f, fs.ErrorIsFile
 		}
 	}
```

Apparently most of the filescom backend code is fine with `f.root = "."` so I'm not sure exactly which bit of the code went wrong. It looks like the `.` got escaped in the list routine to its unicode equivalent

    2024/08/15 11:40:07 DEBUG : GET /api/rest/v1/folders/%EF%BC%8E HTTP/1.1

Which returned

    2024/08/15 11:40:07 DEBUG : HTTP/2.0 404 Not Found

I should really add another integration test to cover this but I had a quick look but the FileIsFile test isn't running in the root so won't trigger the problem.

**Comment by ncw** [comment]
@childish-sambino I think this is ready to merge now. Would you agree?

Alas there is a conflict, so can you rebase it on master first please.

After that I'll merge if you agree.

Thanks

Nick

**Comment by childish-sambino** [comment]
I agree. Ready to merge.

I'm not seeing a conflict.

**Comment by ncw** [comment]
There is a conflict in `fstest/test_all/config.yaml` - don't worry I'll fix this as I have the branch locally already and squash the commits ready for merge.

**Comment by childish-sambino** [comment]
Okay. Cool. Thought I fixed that already.

**Comment by ncw** [comment]
Merged - thank you :-)

**Pull request merged**
